### DISCUSSION TITLE
Nitro layers

### DISCUSSION THREAD

**Discussion by Exotelis** [nitrojs/nitro#2784] [Help Wanted]
After https://github.com/unjs/nitro/pull/2508 was closed and I actually wanted to use modules to add endpoints to an existing BFF, @pi0 pointed out, that layers are a better alternative to modules. These are supposed to work similarly to [Nuxt layers](https://nuxt.com/docs/guide/going-further/layers#basic-example).

However, after some initial attempts, I immediately noticed differences in the behavior of Nitro layers compared to [Nuxt layers](https://nuxt.com/docs/guide/going-further/layers#basic-example). And since layers in Nitro are still a hidden feature (there’s no documentation for them), I’ve only been able to experiment so far.

What doesn’t work are auto imports - both the auto import of functions as well as the auto import of routes. When I use my layer app in a host app, the endpoints simply aren’t present. The runtimeConfig, however, is applied correctly. I created an example on [Stackblitz](https://stackblitz.com/edit/unjs-nitro-starter-fuvf9c?file=README.md).

After going through the docs, types and the source code, I found out that you can solve the problem using handlers in the nitro.config. However, this of course means losing one of the best features - ideally, I wouldn’t want to define routes manually. Additionally, you have to resolve the paths accordingly. The functions such as `defineEventHandler`, or `useRuntimeConfig` must be imported from `#imports`. I also prepared this as an [example](https://stackblitz.com/edit/unjs-nitro-starter-wrxopt?file=README.md).

Is there already a solution for this, can I define a resolver to automatically add handlers? Ofc I could manually parse the src folder, but I would have to do this in any layer I add.

Also, it would be great to have a utility function that retrieves all the defined routes in the host app, or even another layer. For instance, if layerA adds the route `/layerA` and layerB adds `/layerB`, it would be fantastic to generate a list of all the routes.

**Comment by binochoi** [comment]
I give you respect for your adventure.

At this point, scanServerRoutes may be the answer, but it has a deprecated flag and only exists at build or runtime. The function cannot be called in the config file.

Plus at this point nitro's layer doesn't register all plugins, middleware, and routes into the app system. In a nutshell, it doesn't work.

It doesn't seem to have been added to the roadmap. I hope it develops someday.

I think if nitro currently only has the capabilities for layer and dependency infusion, it will be mature enough to switch from nestjs.

**Reply by binochoi** [reply]
For the temporary use of layer, I created a library that add layer's route to the handler list.
If someone is wandering around on a topic related to this, it might help.

https://github.com/binochoi/nitro-extends-routing
